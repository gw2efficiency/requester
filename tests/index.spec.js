/* eslint-env jest */
const fetch = require('../src/index.js')
const fetchMock = require('fetch-mock').sandbox()
const loggerMock = jest.fn()

fetchMock.config.overwriteRoutes = false

beforeEach(() => {
  fetchMock.restore()
  loggerMock.mockReset()

  fetch.logger(loggerMock)
})

function mockResponses (array) {
  array.map(args => {
    fetchMock.mock.apply(fetchMock, args)
  })

  fetch.__set__('fetch', fetchMock)
}

function getLogs () {
  return loggerMock.mock.calls.map(x => ({
    ...x[0],
    duration: Math.round(x[0].duration / 1000) * 1000
  }))
}

describe('requesting', () => {
  it('requests a single url as json', async () => {
    mockResponses([
      ['http://test.com/test', { id: 123 }]
    ])

    let content = await fetch.single('http://test.com/test')
    expect(content).toEqual({ id: 123 })
    expect(getLogs()).toMatchSnapshot()
  })

  it('requests a single url as text', async () => {
    mockResponses([
      ['http://test.com/test', '<h1>Foo</h1>']
    ])

    let content = await fetch.single('http://test.com/test', { type: 'text' })
    expect(content).toEqual('<h1>Foo</h1>')
    expect(getLogs()).toMatchSnapshot()
  })

  it('requests a single url as response object', async () => {
    mockResponses([
      ['http://test.com/test', '<h1>Foo</h1>']
    ])

    let content = await fetch.single('http://test.com/test', { type: 'response' })
    expect(content.url).toEqual('http://test.com/test')
    expect(content.status).toEqual(200)
    expect(content.headers).not.toEqual(undefined)
    expect(getLogs()).toMatchSnapshot()
  })

  it('requests multiple urls as json', async () => {
    mockResponses([
      ['http://test.com/test', { id: 123 }],
      ['http://test.com/test2', { id: 456 }],
      ['http://test.com/test3', { id: 789 }]
    ])

    let content = await fetch.many([
      'http://test.com/test',
      'http://test.com/test2',
      'http://test.com/test3'
    ])
    expect(content).toEqual([{ id: 123 }, { id: 456 }, { id: 789 }])
    expect(getLogs()).toMatchSnapshot()
  })

  it('requests multiple urls as text', async () => {
    mockResponses([
      ['http://test.com/test', '<h1>Foo</h1>'],
      ['http://test.com/test2', '<h1>Foo</h1>'],
      ['http://test.com/test3', '<h1>FooBar</h1>']
    ])

    let content = await fetch.many([
      'http://test.com/test',
      'http://test.com/test2',
      'http://test.com/test3'
    ], { type: 'text' })
    expect(content).toEqual([
      '<h1>Foo</h1>',
      '<h1>Foo</h1>',
      '<h1>FooBar</h1>'
    ])
    expect(getLogs()).toMatchSnapshot()
  })

  it('requests multiple urls as response objects', async () => {
    mockResponses([
      ['http://test.com/test', '<h1>Foo</h1>'],
      ['http://test.com/test2', '<h1>Foo</h1>'],
      ['http://test.com/test3', '<h1>FooBar</h1>']
    ])

    let content = await fetch.many([
      'http://test.com/test',
      'http://test.com/test2',
      'http://test.com/test3'
    ], { type: 'response' })
    expect(content[0].url).toEqual('http://test.com/test')
    expect(content[0].status).toEqual(200)
    expect(content[0].headers).not.toEqual(undefined)
    expect(getLogs()).toMatchSnapshot()
  })
})

describe('waiting', () => {
  it('uses parallel calls if no wait time is specified', async () => {
    mockResponses([
      ['*', () => ({ time: new Date().getTime() })]
    ])

    let timestamps = await fetch.many(['http://1.com', 'http://2.com', 'http://3.com'])

    timestamps = timestamps.map(x => x.time)
    expect(timestamps[1] - timestamps[0]).toBeLessThan(20)
    expect(timestamps[2] - timestamps[1]).toBeLessThan(20)
    expect(getLogs()).toMatchSnapshot()
  })

  it('uses sequential calls and waits if a wait time is specified', async () => {
    mockResponses([
      ['*', () => ({ time: new Date().getTime() })]
    ])

    let timestamps = await fetch.many(
      ['http://1.com', 'http://2.com', 'http://3.com'],
      { waitTime: 100 }
    )

    timestamps = timestamps.map(x => x.time)
    expect(timestamps[1] - timestamps[0]).toBeGreaterThan(99)
    expect(timestamps[2] - timestamps[1]).toBeGreaterThan(99)
    expect(getLogs()).toMatchSnapshot()
  })
})

describe('underlying fetch api', () => {
  it('provides "fetch" with the options', async () => {
    mockResponses([
      ['http://test.com/test', { id: 123 }]
    ])

    let content = await fetch.single('http://test.com/test')
    expect(content).toEqual({ id: 123 })
    expect(fetchMock.lastOptions()).toEqual({
      type: 'json',
      method: 'GET',
      headers: {},
      body: undefined
    })
    expect(getLogs()).toMatchSnapshot()
  })

  it('can overwrite the default "fetch" options', async () => {
    mockResponses([
      ['http://test.com/test', '<h1>Foo</h1>']
    ])

    let options = {
      type: 'text',
      method: 'POST',
      headers: { 'Authentication': 'Test' },
      body: 'foo=bar'
    }
    let content = await fetch.single('http://test.com/test', options)
    expect(content).toEqual('<h1>Foo</h1>')
    expect(fetchMock.lastOptions()).toEqual(options)
    expect(getLogs()).toMatchSnapshot()
  })
})

describe('error handling', () => {
  it('throws an error if a request fails', async () => {
    mockResponses([
      ['http://failing.com/yes', 500]
    ])

    try {
      await fetch.single('http://failing.com/yes')
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('throws an error if a request fails even when we get the response object', async () => {
    mockResponses([
      ['http://failing.com/yes', 500]
    ])

    try {
      await fetch.single('http://failing.com/yes', { type: 'response' })
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('throws an error if a request of many fails', async () => {
    mockResponses([
      ['http://failing.com/no', '<h1>Foo</h1>'],
      ['http://failing.com/yes', 500]
    ])

    try {
      await fetch.many([
        'http://failing.com/no',
        'http://failing.com/yes'
      ], { type: 'text' })
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('throws an decoding exception for malformed json', async () => {
    mockResponses([
      ['http://failing.com/malformed', '{"test: "malformed"}']
    ])

    try {
      await fetch.single('http://failing.com/malformed')
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)
    expect(err.message).toEqual(expect.stringContaining('Unexpected token'))
    expect(getLogs()).toMatchSnapshot()
  })

  it('always throws an status exception from a bad status', async () => {
    mockResponses([
      ['http://failing.com/malformed', { status: 500, body: 'Error message which is not JSON' }]
    ])

    try {
      await fetch.single('http://failing.com/malformed')
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)
    expect(err.message).toEqual('Status 500')
    expect(getLogs()).toMatchSnapshot()
  })

  it('throws an error that includes the response and the content', async () => {
    mockResponses([
      ['http://failing.com/erroring', { status: 403, body: '{"text": "authentication required"}' }]
    ])

    try {
      await fetch.single('http://failing.com/erroring')
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)
    expect(typeof err.response).toEqual('object')
    expect(err.message).toEqual('Status 403')
    expect(err.response.status).toEqual(403)
    expect(err.content).toEqual({ text: 'authentication required' })
    expect(getLogs()).toMatchSnapshot()
  })
})

describe('retrying', () => {
  it('retries on error till success', async () => {
    fetch.retry(() => true)

    mockResponses([
      ['http://test.com/test', 500, { repeat: 3 }],
      ['http://test.com/test', 'text', { repeat: 1 }],
      ['http://test.com/test', 'not text']
    ])

    let content = await fetch.single('http://test.com/test', { type: 'text' })
    expect(content).toEqual('text')
    expect(getLogs()).toMatchSnapshot()
  })

  it('respects the decider response', async () => {
    fetch.retry((tries) => tries <= 3)

    mockResponses([
      ['http://test.com/test', 500, { repeat: 3 }],
      ['http://test.com/test', 'text', { repeat: 1 }],
      ['http://test.com/test', 'not text']
    ])

    try {
      await fetch.single('http://test.com/test')
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('can disable retrying', async () => {
    fetch.retry(() => false)

    mockResponses([
      ['http://test.com/test', 500, { repeat: 1 }],
      ['http://test.com/test', 'text', { repeat: 1 }],
      ['http://test.com/test', 'not text']
    ])

    try {
      await fetch.single('http://test.com/test')
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)
    expect(getLogs()).toMatchSnapshot()
  })

  it('can access the error object and the retries in the retry decider', async () => {
    let callbackMock = jest.fn()
    fetch.retry(callbackMock)

    mockResponses([
      ['http://test.com/test', 500, { repeat: 1 }],
      ['http://test.com/test', 'text', { repeat: 1 }],
      ['http://test.com/test', 'not text']
    ])

    try {
      await fetch.single('http://test.com/test')
    } catch (e) {
      var err = e
    }

    expect(err).toBeInstanceOf(Error)

    let deciderArguments = callbackMock.mock.calls[0]
    expect(deciderArguments[0]).toEqual(2)
    expect(deciderArguments[1]).toBeInstanceOf(Error)
    expect(deciderArguments[1].response).not.toEqual(undefined)
    expect(deciderArguments[1].response.status).toEqual(500)
    expect(getLogs()).toMatchSnapshot()
  })

  it('can specify a wait function for retrying', async () => {
    fetch.retry(() => true)
    fetch.retryWait(tries => tries * 100)

    mockResponses([
      ['http://test.com/test', 500, { repeat: 4 }],
      ['http://test.com/test', 'text', { repeat: 1 }],
      ['http://test.com/test', 'not text']
    ])

    let start = new Date()
    await fetch.single('http://test.com/test', { type: 'text' })
    expect(new Date() - start).toBeGreaterThan(99 * (1 + 2 + 3 + 4))
    expect(getLogs()).toMatchSnapshot()
  })
})
