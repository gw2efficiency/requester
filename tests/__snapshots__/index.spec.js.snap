// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error handling always throws an status exception from a bad status 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://failing.com/malformed",
  },
]
`;

exports[`error handling throws an decoding exception for malformed json 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://failing.com/malformed",
  },
]
`;

exports[`error handling throws an error if a request fails 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://failing.com/yes",
  },
]
`;

exports[`error handling throws an error if a request fails even when we get the response object 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://failing.com/yes",
  },
]
`;

exports[`error handling throws an error if a request of many fails 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://failing.com/no",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://failing.com/yes",
  },
]
`;

exports[`error handling throws an error that includes the response and the content 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 403,
    "url": "http://failing.com/erroring",
  },
]
`;

exports[`requesting requests a single url as json 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test",
  },
]
`;

exports[`requesting requests a single url as response object 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test",
  },
]
`;

exports[`requesting requests a single url as text 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test",
  },
]
`;

exports[`requesting requests multiple urls as json 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test2",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test3",
  },
]
`;

exports[`requesting requests multiple urls as response objects 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test2",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test3",
  },
]
`;

exports[`requesting requests multiple urls as text 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test2",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test3",
  },
]
`;

exports[`retrying can access the error object and the retries in the retry decider 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://test.com/test",
  },
]
`;

exports[`retrying can disable retrying 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://test.com/test",
  },
]
`;

exports[`retrying can specify a wait function for retrying 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 1,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 2,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 3,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 4,
    "status": 200,
    "url": "http://test.com/test",
  },
]
`;

exports[`retrying respects the decider response 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 1,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 2,
    "status": 500,
    "url": "http://test.com/test",
  },
]
`;

exports[`retrying retries on error till success 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 1,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 2,
    "status": 500,
    "url": "http://test.com/test",
  },
  Object {
    "duration": 0,
    "retries": 3,
    "status": 200,
    "url": "http://test.com/test",
  },
]
`;

exports[`underlying fetch api can overwrite the default "fetch" options 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test",
  },
]
`;

exports[`underlying fetch api provides "fetch" with the options 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://test.com/test",
  },
]
`;

exports[`waiting uses parallel calls if no wait time is specified 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://1.com",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://2.com",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://3.com",
  },
]
`;

exports[`waiting uses sequential calls and waits if a wait time is specified 1`] = `
Array [
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://1.com",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://2.com",
  },
  Object {
    "duration": 0,
    "retries": 0,
    "status": 200,
    "url": "http://3.com",
  },
]
`;
